File: Services/DailyReflectionService.swift
import Foundation

struct DailyReflectionService {
    func reflectionForToday(from entries: [ReflectionEntry], preferredSources: [SourceType]) -> ReflectionEntry? {
        let filtered = entries.filter { preferredSources.contains($0.sourceType) }
        guard !filtered.isEmpty else { return entries.first }

        let today = Calendar.current.ordinality(of: .day, in: .year, for: Date()) ?? 1
        let index = (today - 1) % filtered.count
        return filtered.sorted(by: { $0.id < $1.id })[index]
    }
}

File: GurmukhiFlutter/lib/services/daily_reflection_service.dart
import '../models/reflection_entry.dart';
import '../models/source_type.dart';

class DailyReflectionService {
  ReflectionEntry? reflectionForToday(
    List<ReflectionEntry> entries,
    List<SourceType> preferredSources,
  ) {
    final filtered = entries.where((entry) => preferredSources.contains(entry.sourceType)).toList()
      ..sort((a, b) => a.id.compareTo(b.id));

    if (filtered.isEmpty) {
      return entries.isEmpty ? null : entries.first;
    }

    final dayOfYear = DateTime.now().difference(DateTime(DateTime.now().year, 1, 1)).inDays + 1;
    final index = (dayOfYear - 1) % filtered.length;
    return filtered[index];
  }
}
